src/main/scala/uk/gov/nationalarchives/consignmentexport/Main.scala

```diff
@@ -25,20 +25,27 @@ object Main extends CommandIOApp("tdr-consignment-export", "Exports tdr files in
         rootLocation = config.efs.rootLocation
         exportId = UUID.randomUUID
         basePath = s"$rootLocation/$exportId"
-        tarPath = s"${basePath}/$consignmentId.tar.gz"
+        tarPath = s"$basePath/$consignmentId.tar.gz"
         bashCommands = BashCommands()
         graphQlApi = GraphQlApi(config.api.url, config.auth.url)
         keycloakClient = KeycloakClient(config)
         s3Files = S3Files(S3Utils(s3Async))
+        bagit = Bagit()
+        validator = Validator(consignmentId)
         //Export datetime generated as value needed in bag metadata and DB table
         //Cannot use the value from DB table in bag metadata, as bag metadata created before bagging
         //and cannot update DB until bag creation successfully completed
         exportDatetime = ZonedDateTime.now(ZoneOffset.UTC)
-        bagMetadata <- BagMetadata(graphQlApi, keycloakClient).getBagMetadata(consignmentId, config, exportDatetime)
-        data <- graphQlApi.getFiles(config, consignmentId)
-        _ <- IO.fromOption(data.headOption)(new Exception(s"Consignment API returned no files for consignment $consignmentId"))
-        _ <- s3Files.downloadFiles(data, config.s3.cleanBucket, consignmentId, basePath)
-        _ <- Bagit().createBag(consignmentId, basePath, bagMetadata)
+        consignmentResult <- graphQlApi.getConsignmentMetadata(config, consignmentId)
+        consignmentData <- validator.validateConsignmentResult(consignmentResult)
+        _ <- validator.validateConsignmentHasFiles(consignmentData)
+        bagMetadata <- BagMetadata(keycloakClient).generateMetadata(consignmentId, consignmentData, exportDatetime)
+        validatedFileMetadata <- validator.validateFileMetadataNotEmpty(consignmentData.files)
+        _ <- s3Files.downloadFiles(validatedFileMetadata, config.s3.cleanBucket, consignmentId, basePath)
+        bag <- bagit.createBag(consignmentId, basePath, bagMetadata)
+        fileMetadataCsv <- BagAdditionalFiles(bag.getRootDir).fileMetadataCsv(validatedFileMetadata)
+        checksums <- ChecksumCalculator().calculateChecksums(fileMetadataCsv)
```

Comment:
Does Bagit not provide a way to do this? I'm just surprised we have to generate our own SHA-256 when checksums and tag files are a standard part of Bagit.

Reply:
I don't think so, if you look at the `gov.loc.repository.bagit.domain.Manifest` class which is the class you write out when you're writing to the tag file, there's a `fileToChecksumMap` property. I imagine it's because `file-metadata.csv` is a non-standard file that's not in the Bagit spec and so it doesn't have any provision to calculate checksums for it but the library is quite happy for you to provide your own.